File: frontend/ios/Hive/Hive/PostFeedManager.swift
//
//  PostFeedManager.swift
//  Hive
//
//  Created by Chuck Onwuzuruike on 9/22/19.
//  Copyright © 2019 Chuck Onwuzuruike. All rights reserved.
//

import Foundation
import UIKit

protocol PostFeedDelegate: class {
    func showComments(postView: PostView)
    func fetchMorePosts(queryParams: QueryParams)
    func performAction(post: Post, actionType: ActionType)
}

class PostFeedManager: NSObject, PostViewDelegate {
    
    let POST_VIEW_CELL_REUSE_IDENTIFIER = "postView"
    let POST_VIEW_CELL_NIB_NAME = "PostView"
    
    private weak var _postTableView: UITableView?
    public private(set) var posts: Array<Post> = []
    public private(set) var delegate: PostFeedDelegate?
    private var prevFetchQueryMetadata: QueryMetadata = QueryMetadata()
    private var refreshControl: UIRefreshControl?
    
    var postTableView: UITableView {
        get { return _postTableView! }
    }

    func configure(tableView: UITableView, delegate: PostFeedDelegate) {
        self._postTableView = tableView
        self.delegate = delegate
        // Register nib as cell.
        // Overrides whatever identifier is used in storyboard... I think.
        postTableView.register(UINib(nibName: POST_VIEW_CELL_NIB_NAME, bundle: nil),
                               forCellReuseIdentifier: POST_VIEW_CELL_REUSE_IDENTIFIER)
        postTableView.delegate = self
        postTableView.dataSource = self
        postTableView.isScrollEnabled = true
        postTableView.refreshControl = UIRefreshControl()
        refreshControl = postTableView.refreshControl
        refreshControl!.addTarget(self, action: #selector(refreshPostTableView(_:)), for: .valueChanged)
        refreshControl!.attributedTitle = NSAttributedString(string: "ugh, hold on...")
        
        // Fetch an initial set of hosts.
        fetchMorePosts(getNewer: true)
    }
    
    // What to do when user forces a refresh on the table
    @objc func refreshPostTableView(_ refreshControl: UIRefreshControl) {
        print("fetch newer posts")
        fetchMorePosts(getNewer: true)
    }
    
    func reloadUI() {
        DispatchQueue.main.async {
            self.postTableView.reloadData()
        }
        setRefreshing(set: false)
    }
    
    func setRefreshing(set: Bool) {
        DispatchQueue.main.async {
            if (self.refreshControl != nil) {
                if (set && !self.refreshControl!.isRefreshing) {
                    print("Start PostFeed Refreshing...")
                    self.refreshControl?.beginRefreshing()
                } else {
                    print("Stop PostFeed Refreshing...")
                    self.refreshControl?.endRefreshing()
                }
            }
        }
    }
    
    // Called by Controller to give us more hosts.
    func addMorePosts(morePosts: Array<Post>, newMetadata: QueryMetadata) {
        self.prevFetchQueryMetadata.updateMetadata(newMetadata: newMetadata)
        var set: Set = Set(morePosts)
        
        // Merge old posts and new (morePosts) posts. If a post exists in old and in new, take the
        // new post.
        for post in posts {
            if !set.contains(post) {
                set.insert(post)
            }
        }
        posts = Array(set)
            .filter {!$0.isExpired()}
            .sorted(by: {$0.creationTimestampSec > $1.creationTimestampSec})
    }
    
    func resetData() {
        self.posts = []
        self.prevFetchQueryMetadata = QueryMetadata()
    }
    
    func reset(newTableView: UITableView) {
        self._postTableView = newTableView
        resetData()
    }
    
    // A poke to the manager to ask for more new posts
    func pokeNew() {
        fetchMorePosts(getNewer: true)
    }
    
    // Will reset data and then request new data
    func resetDataAndPokeNew() {
        resetData()
        pokeNew()
    }
    
    func reconfigureWithAction(postId: String, actionType: ActionType) {
        let index = posts.firstIndex(where: {$0.postId == postId})
        if index == nil {
            return;
        }
        let i = index!
        // filter will return a list but it should be one item long, unless
        // the post doesn't exist anymore or some weird duplicate error, which
        // shouldn't happen since we remove duplicates.

        // Now change the # of likes and dislikes by 1 depending on the old and
        // new action type.
        let newActionType = actionType
        let oldActionType = posts[i].userActionType
        if oldActionType == ActionType.LIKE {
            switch newActionType {
            case ActionType.LIKE:
                posts[i].likes -= 1;
                break;
            case ActionType.DISLIKE:
                posts[i].likes -= 1;
                posts[i].dislikes += 1;
                break;
            case ActionType.NO_ACTION:
                posts[i].likes -= 1;
                break;
            }
        }
        if oldActionType == ActionType.DISLIKE {
            switch newActionType {
            case ActionType.LIKE:
                posts[i].dislikes -= 1;
                posts[i].likes += 1;
                break;
            case ActionType.DISLIKE:
                posts[i].dislikes -= 1;
                break;
            case ActionType.NO_ACTION:
                posts[i].dislikes -= 1;
                break;
            }
        }
        if oldActionType == ActionType.NO_ACTION {
            switch newActionType {
            case ActionType.LIKE:
                posts[i].likes += 1;
                break;
            case ActionType.DISLIKE:
                posts[i].dislikes += 1;
                break;
            case ActionType.NO_ACTION:
                print("WARNING: This shouldn't happen but its not really an error.")
                break;
            }
        }
        posts[i].userActionType = newActionType
        reloadUI()
    }
    
    private func fetchMorePosts(getNewer: Bool) {
        if (!getNewer && !(prevFetchQueryMetadata.hasMoreOlderData ?? true)) {
            print("Server already told us there are no more old posts. Returning.")
            return
        }
        if (getNewer) {
            setRefreshing(set: true)
        }
        let params: QueryParams = QueryParams(getNewer: getNewer,
                                              currTopCursorStr: self.prevFetchQueryMetadata.newTopCursorStr,
                                              currBottomCursorStr: self.prevFetchQueryMetadata.newBottomCursorStr)
        delegate?.fetchMorePosts(queryParams: params)
    }
    
    func commentButtonClick(postView: PostView) {
        delegate?.showComments(postView: postView);
    }
    
    func performAction(postView: PostView, actionType: ActionType) {
        delegate?.performAction(post: postView.post, actionType: actionType)
    }
}

extension PostFeedManager: UITableViewDelegate {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }

    // Asks for cell to display at indexPath
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: POST_VIEW_CELL_REUSE_IDENTIFIER) as! PostView
        cell.configure(post: posts[indexPath.section], delegate: self)
        cell.layer.borderWidth = 2
        cell.layer.shadowColor = UIColor.black.cgColor
        cell.layer.shadowOpacity = 0.5
        cell.layer.shadowOffset = CGSize(width: 0.0, height: 2.0)
        cell.layer.borderColor = UIColor.white.cgColor
        return cell
    }
}

extension PostFeedManager: UITableViewDataSource {
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 15
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView?
    {
        let headerView = UIView(frame: CGRect(x: 0, y: 0, width: tableView.bounds.size.width, height: 10))
        headerView.backgroundColor = UIColor.clear
        
        return headerView
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 120
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.posts.count
    }
    
    // Gives me cell from cellForRowAt right before it is displayed.
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        if indexPath.section == tableView.numberOfSections - 1 {
            if (!posts.isEmpty) {
                print("fetch older posts")
                fetchMorePosts(getNewer: false)
            }
        }
    }
}

File: frontend/android/app/src/main/java/coloredcoded/hive/PostFeedManager.java
package coloredcoded.hive;

import android.content.Context;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.Iterator;

import coloredcoded.hive.client.ActionType;
import coloredcoded.hive.client.Post;
import coloredcoded.hive.client.QueryMetadata;
import coloredcoded.hive.client.QueryParams;

// Manages all the things having to do with interacting with a post feed. Essentially, is managing
// a list view.
public class PostFeedManager implements PostView.Delegate {

    // Will delegate tasks to the fragment that created it.
    public interface Delegate {
        void showComments(PostView postView);
        void fetchMorePosts(QueryParams queryParams);
        void performAction(Post post, ActionType actionType);
    }

    // Adapter in charge of making sure that the post feed displays correctly. PostFeedManager could
    // subclass ArrayAdapter directly. Perhaps we should consider doing that.
    private class PostFeedAdapter extends ArrayAdapter<Post> {

        private PostView.Delegate postViewDelegate;

        public PostFeedAdapter(Context context, PostView.Delegate delegate,
                               ArrayList<Post> thePosts) {
            super(context, -1, thePosts);
            postViewDelegate = delegate;
        }

        @NonNull
        @Override
        public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
            PostView pv = PostView.newInstance(getItem(position), postViewDelegate, parent);
            if (disableLikeAndDislikeButtons) {
                pv.disableLikeAndDislikeButtons();
            }
            return pv.getPostView();
        }
    }

    private ArrayList<Post> posts;
    private QueryMetadata prevQueryMetadata;
    private ArrayAdapter<Post> postFeedAdapter;
    private ListView postFeedListView;
    private SwipeRefreshLayout swipeRefreshLayout;
    private Delegate delegate;
    private boolean disableLikeAndDislikeButtons;

    public PostFeedManager(Context context) {
        posts = new ArrayList<>();
        prevQueryMetadata = new QueryMetadata();
        postFeedAdapter = new PostFeedAdapter(context, this, posts);
        disableLikeAndDislikeButtons = false;
    }

    public void configure(ListView listView, SwipeRefreshLayout refreshLayout, Delegate delegate) {
        postFeedListView = listView;
        this.delegate = delegate;
        postFeedListView.setAdapter(postFeedAdapter);
        swipeRefreshLayout = refreshLayout;

        setupListeners();
        fetchMorePosts(true);
    }

    public Post findPostById(String postId) {
        Post post = null;
        Iterator<Post> itr = posts.iterator();
        while (itr.hasNext()) {
            post = itr.next();
            if (post.getPostId().equals(postId)) {
                break;
            }
        }
        return post;
    }

    private void setupListeners() {
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                System.out.println("Refreshing..");
                fetchMorePosts(true);
            }
        });
        postFeedListView.setOnScrollListener(new AbsListView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(AbsListView view, int scrollState) {
            }

            @Override
            public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount,
                                 int totalItemCount) {
                if (swipeRefreshLayout.isRefreshing()) {
                    return;
                }
                if (posts.isEmpty()) {
                    return;
                }
                // If we are displaying the last set of items in the list view.
                if (firstVisibleItem + visibleItemCount >= totalItemCount -1) {
                    fetchMorePosts(false);
                }
            }
        });
    }

    public void setRefreshAnimation(final boolean set) {
        swipeRefreshLayout.post(new Runnable() {
            @Override
            public void run() {
                swipeRefreshLayout.setRefreshing(set);
            }
        });
    }

    private void fetchMorePosts(boolean getNewer) {
        System.out.println("Fetching " + (getNewer ? "newer" : "older") + " posts");
        if (!getNewer && !prevQueryMetadata.hasMoreOlderData()) {
            System.out.println("Server already told us there are no more old posts. Returning.");
            return;
        }
        if (getNewer) {
            setRefreshAnimation(true);
        }
        QueryParams params = new QueryParams(getNewer,
                prevQueryMetadata.getNewTopCursorStr(),
                prevQueryMetadata.getNewBottomCursorStr());
        delegate.fetchMorePosts(params);
    }

    // Updates the list view. Must be called on the UI thread.
    public void reloadUI() {
        postFeedListView.post(
                new Runnable() {
                    @Override
                    public void run() {
                        postFeedAdapter.notifyDataSetChanged();
                        setRefreshAnimation(false);
                    }
                }
        );
    }

    public void addMorePosts(ArrayList<Post> morePosts, QueryMetadata newMetadata) {
        prevQueryMetadata.updateMetadata(newMetadata);

        HashSet<Post> set = new HashSet<>(morePosts);

        // Merge old posts and new (morePosts) posts. If a post exists in old and in new, take the
        // new post.
        for (Post p : posts) {
            if (!set.contains(p)) {
                set.add(p);
            }
        }

        // Don't include expired hosts while transferring from set back to posts.
        Iterator<Post> itr = set.iterator();
        posts.clear();
        while (itr.hasNext()) {
            Post p = itr.next();
            if (!p.isExpired()) {
                posts.add(p);
            }
        }

        // Sort posts.
        Collections.sort(posts, new Comparator<Post>() {
            @Override
            public int compare(Post o1, Post o2) {
                return (int) (o2.getCreationTimestampSec() -
                        o1.getCreationTimestampSec());
            }
        });

        reloadUI();
    }

    public void incrementNumberOfComments(String postId) {
        findPostById(postId).incrementNumberOfComments();
    }

    public void updateActionType(String postId, ActionType actionType) {
        // Find post with id.
        Post post = findPostById(postId);
        Iterator<Post> itr = posts.iterator();

        // Now change the # of likes and dislikes by 1 depending on the old and
        // new action type.
        ActionType newActionType = actionType;
        ActionType oldActionType = post.getUserActionType();
        if (oldActionType == ActionType.LIKE) {
            post.addLikes(-1);
            if (newActionType == ActionType.DISLIKE) {
                post.addDislikes(1);
            }
        }
        if (oldActionType == ActionType.DISLIKE) {
            post.addDislikes(-1);
            if (newActionType == ActionType.LIKE) {
                post.addLikes(1);
            }
        }
        if (oldActionType == ActionType.NO_ACTION) {
            switch (newActionType) {
                case LIKE:
                    post.addLikes(1);
                    break;
                case DISLIKE:
                    post.addDislikes(1);
                    break;
                case NO_ACTION:
                    Log.e("updateActionType",
                            "WARNING: This shouldn't happen but its not really an error.");
                    break;
            }
        }
        post.setUserActionType(newActionType);
        reloadUI();
    }

    public void resetData() {
        posts.clear();
        prevQueryMetadata = new QueryMetadata();
    }

    public void pokeNew() {
        fetchMorePosts(true);
    }

    public void resetDataAndPokeNew() {
        System.out.println("Attempting to reset data");
        resetData();
        pokeNew();
    }

    public void setDisableLikeAndDislikeButtons(boolean set) {
        disableLikeAndDislikeButtons = set;
    }

    @Override
    public void commentButtonClick(PostView postView) {
        delegate.showComments(postView);
    }

    @Override
    public void performAction(PostView postView, ActionType actionType) {
        delegate.performAction(postView.getPost(), actionType);
    }
}
